source/load-balancers/tests/transactions.yml

```diff
@@ -0,0 +1,473 @@
+          - commandStartedEvent:
+              commandName: killCursors
+          - commandStartedEvent:
+              commandName: commitTransaction
+      - client: *client0
+        eventType: cmap
+        events:
+          # Only one set of events should be published because all of the operations use the connection pinned by the
+          # transaction.
+          - connectionReadyEvent: {}
+          - connectionCheckedOutEvent: {}
+          - connectionCheckedInEvent: {}
+
+  - description: pinned connections are returned after a network error on a CRUD command
+    operations:
+      - name: failPoint
+        object: testRunner
+        arguments:
+          client: *client0
+          failPoint:
+            configureFailPoint: failCommand
+            mode: { times: 1 }
+            data:
+              failCommands: [update]
+              closeConnection: true
+      - name: startTransaction
+        object: *session0
+      - name: insertOne
+        object: *collection0
+        arguments:
+          document: *insertDocument
+          session: *session0
+      - name: assertNumberConnectionsCheckedOut
+        object: testRunner
+        arguments:
+          client: *client0
+          connections: 1
+      - name: updateOne
+        object: *collection0
+        arguments:
+          filter: *insertDocument
+          update:
+            $set:
+              x: 1
+          session: *session0
+        expectError:
+          # Network errors are considered client-side errors by the unified test format.
+          isClientError: true
+      - name: assertNumberConnectionsCheckedOut
+        object: testRunner
+        arguments:
+          client: *client0
+          connections: 0
+      - name: abortTransaction
+        object: *session0
+    expectEvents:
+      - client: *client0
+        events:
+          - commandStartedEvent:
+              command:
+                insert: *collection0Name
+                documents: [*insertDocument]
+              commandName: insert
+          - commandStartedEvent:
+              command:
+                update: *collection0Name
+              commandName: update
+          - commandStartedEvent:
+              commandName: abortTransaction
+      - client: *client0
+        eventType: cmap
+        events:
+          # Events for setting the failpoint.
+          - connectionReadyEvent: {}
+          - connectionCheckedOutEvent: {}
+          - connectionCheckedInEvent: {}
+          # Events to run the insert/update commands.
+          - connectionCheckedOutEvent: {}
+          - connectionCheckedInEvent: {}
+          - connectionClosedEvent:
+              reason: error
+          # Events for the abortTransaction attempt.
+          - connectionReadyEvent: {}
+          - connectionCheckedOutEvent: {}
+          - connectionCheckedInEvent: {}
+
+  # This is similar to the previous test, but it asserts that the connection pinned by the transaction is returned to
```

Comment:
I'm not sure that this is a reasonable thing for drivers to do.  I think we should consider whether we will require the cursor to be closed in order for it to unpin its connection.

If we decide that this behavior is what's desired, it's going to be tricky to implement in drivers that are using reference counting idioms to track when to return a connection to the pool.  Java driver certainly has this problem.

For now, I was able to get this test passing by adding a `close() operation after the failed `updateOne`.